File: arrays-hashing/49_group_anagrams.py
from collections import defaultdict
from typing import Dict, List, Tuple

class Solution:
    def groupAnagrams(self, strs: List[str]) -> List[List[str]]:
        groups: Dict[Tuple[int], List[str]] = defaultdict(list)
        for word in strs:
            counts: List[int] = [0 for _ in range(26)]
            for char in word:
                counts[ord(char) - ord('a')] += 1
            groups[tuple(counts)].append(word)
        return groups.values()

File: arrays-hashing/49_group_anagrams.cpp
#include <algorithm>
#include <vector>
#include <string>
#include <unordered_map>
#include <vector>

class Solution {
public:
    std::vector<std::vector<std::string>> groupAnagrams(std::vector<std::string>& strs) {
        std::unordered_map<std::string, std::vector<std::string>> group_map {};
        for (const std::string& str : strs) {
            std::string sorted_str = std::string {str};
            std::sort(sorted_str.begin(), sorted_str.end());
            group_map[sorted_str].push_back(str);
        }
        std::vector<std::vector<std::string>> groups;
        for (const auto& [sorted_str, str_groups] : group_map) {
            groups.push_back(str_groups);
        }
        return groups;
    }
};
